### ISSUE TITLE
-Wreturn-stack-address false positive with unreachable code

### ISSUE THREAD

**Issue by Soveu** [llvm/llvm-project#166374]
https://cpp.godbolt.org/z/3357xPfP8

```cpp
#include <cstdlib>

template <typename R>
[[noreturn]] constexpr R __absurd() { std::abort(); }

int const& weird() {
    return __absurd<int>();
}
```

Technically, `weird` should return `__absurd<int const&>()`, but I accidentally wrote it like this and I'm not really sure if this counts as a bug.
